ducktape/mark/_mark.py

```diff
@@ -159,6 +159,8 @@ def apply(self, seed_context, context_list):
                 for injected_args in cartesian_product_dict(
                         {arg: self.injected_args[arg] for arg in self.injected_args if arg not in ctx.injected_args}):
                     injected_args.update(ctx.injected_args)
+                    if ctx.cluster_use_metadata:
+                        seed_context.cluster_use_metadata = ctx.cluster_use_metadata
```

Comment:
Modifying the function argument doesn't look like a good idea, especially since this seed context will be passed to other mark implementations. Can you explain why the fix is the way it is and why do you need to modify seed context in defaults but not in other marks?
(also, you don't do anything in the `else` where context_list is none)

Reply:
I'm not sure that functionality we rely upon so much (the order in which marks are applied and how multiple cluster annotations are applied) is actually by design - but could be wrong

Reply:
Sure, 
In the current implementation, since cluster mark is on top followed by defaults mark, the named attribute we get from object is (say)
```
[<ducktape.mark._mark.Matrix object at 0x7f9a691a8e20>, 
<ducktape.mark.resource.ClusterUseMetadata object at 0x7f9a691a8d30>, 
<ducktape.mark._mark.Parametrize object at 0x7f9a691a8d00>, 
<ducktape.mark._mark.Parametrize object at 0x7f9a691a8cd0>, 
<ducktape.mark.resource.ClusterUseMetadata object at 0x7f9a691a8c70>, 
<ducktape.mark._mark.Parametrize object at 0x7f9a691a8ca0>, 
<ducktape.mark._mark.Defaults object at 0x7f9a691a8c40>, 
<ducktape.mark.resource.ClusterUseMetadata object at 0x7f9a691a8c10>]
```
when we reach to ‘defaults’ in [this for loop here](https://github.com/confluentinc/ducktape/blob/v0.7.12/ducktape/mark/mark_expander.py#L42-L44), we execute [this mark ‘apply’ for ‘default’](https://github.com/confluentinc/ducktape/blob/6a8d9c28cf62227761e4e531c09b4c95b4d3104c/ducktape/mark/_mark.py#L155-L171) piece of code.
Here, we pass the context_list and seed_context
Even though we have our cluster metadata in context_list since this is the ‘default’ case, we define an empty new context list [here](https://github.com/confluentinc/ducktape/blob/6a8d9c28cf62227761e4e531c09b4c95b4d3104c/ducktape/mark/_mark.py#L156), get injected args from the cartesian product, and set the cluster metadata to empty because cluster metadata for `seed_context` is `{}` and insert that into the new context list [here](https://github.com/confluentinc/ducktape/blob/6a8d9c28cf62227761e4e531c09b4c95b4d3104c/ducktape/mark/_mark.py#L165 and return that new context list https://github.com/confluentinc/ducktape/blob/6a8d9c28cf62227761e4e531c09b4c95b4d3104c/ducktape/mark/_mark.py#L171) opposed to updating the initial context text as we do at other places...
Parametrize: https://github.com/confluentinc/ducktape/blob/6a8d9c28cf62227761e4e531c09b4c95b4d3104c/ducktape/mark/_mark.py#L189
Matrix: https://github.com/confluentinc/ducktape/blob/6a8d9c28cf62227761e4e531c09b4c95b4d3104c/ducktape/mark/_mark.py#L130
After this, the returned context list would have empty cluster metadata. In the next iteration, we would be going over `<ducktape.mark.resource.ClusterUseMetadata object at 0x7f9a691a8c10>` and this looks for `@cluster(num_nodes=...)` that is above the default and apply that to the entire context. If `@cluster(num_nodes=...)`  above `@defaults` is not defines, it will take it as empty and use the entire cluster. 

In this PR, 
we check if [our context](https://github.com/confluentinc/ducktape/blob/6a8d9c28cf62227761e4e531c09b4c95b4d3104c/ducktape/mark/_mark.py#L162) has cluster metadata, we use that `ctx.cluster_use_metadata` instead of `{}` and solve our problem. 
we don't need an `else` because if our `ctx.cluster_use_metadata` is `{}`, then it's the same as using seed_context.cluster_use_metadata which is `{}`

Reply:
Yeah, I understand that defaults creates new contexts based on the cartesian product of the context args x default args. And you're right, we don't need to do anything if context_list is empty.

What concerns me is that modifying seed_context doesn't look right.

_(I'll leave aside the the question of why not just do `ctx.copy` instead of `seed_context.copy` - it was like this before, could've been made like that for a reason; or not)_

Why not smth like this?
```python
new_context = seed_context.copy(
    function=injected_fun, 
    injected_args=injected_args,
    cluster_use_metadata=ctx.cluster_use_metadata)
new_context_list.insert(0, new_context)
```

Reply:
Agreed. probably for the best 👍 
Pushed commit with updated changes. Thanks!